File: 17-LetterCombinationsofaPhoneNumber.java
public class Solution {
    public List<String> letterCombinations(String digits) {
        
        List<String> res = new ArrayList<>();
        if (digits.length() == 0)    return res;
        String[] letterMap = {"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        res.add("");
        for (int i = 0; i < digits.length(); i++){
            List<String> temp = new ArrayList<>();
            //int curr = Character.getNumericValue(digits[i]);
            int curr = digits.charAt(i) - '0';
            String chars = letterMap[curr];
            for (int x = 0; x < chars.length(); x++){
                for(int j = 0; j < res.size(); j++){
                    temp.add(res.get(j)+chars.charAt(x));
                }
            }
            res = temp;
        }
        return res;
    }
}

File: cpp/17-LetterCombinationsOfAPhoneNumber.cpp
//Input:Digit string "23"
//Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        string dict[] = {"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        if (digits.empty())
            return res;
        res.push_back("");
        for (int i = 0; i < digits.size(); i++) {
            int currDigit = digits[i] - '0';
            string currLetters = dict[currDigit];
            vector<string> tmp;
            for (int j = 0; j < currLetters.size(); j++) {
                for (int k = 0; k < res.size(); k++) {
                    tmp.push_back(res[k] + currLetters[j]);
                }
            }
            res = tmp;
        }
        return res;
    }
};
